### PR TITLE
[Pegasus] Refactor Tokenizer

### LINKED ISSUE #8594: PEGASUS do not have mask token
@mfuntowicz @patrickvonplaten 

Hi,

I am using PEGASUS  google/pegasus-large

I would like to fill the mask sentence of a document , i.e. try the pretraining task. But I don't find the mask_token. 

1. from transformers import PegasusTokenizer
2. tok = PegasusTokenizer.from_pretrained("google/pegasus-large")
3. tok.mask_token

output is "Using mask_token, but it is not set yet."

### LINKED ISSUE #8689: [Question] Pegasus tokenizer
@sshleifer - sorry to ping you here on this. Would be amazing if you find some time to explain the Pegasus tokenizer a bit.

A couple of things I don't understand:

- In the official Pegasus Tokenizer and from reading the paper it seems that exactly 2 mask tokens are necessary.
    See https://github.com/google-research/pegasus/blob/master/pegasus/ops/pretrain_parsing_ops.cc#L66
   a) ID=2 seems to correspond to the sentence mask token, called `[MASK_1]` and 
   b) ID=3 seems to correspond to the word mask token, called `[MASK_2]`

=> Why don't we have `[MASK_1]` and `[MASK_2]` tokens in the tokenizer's special tokens? I would actually add them at the id's 2 and 3 instead of having `unk_2` and `unk_3` there. Wdyt?

- Why do we call the tokens unk_2 - unk_104 ? Why unk? And why aren't those part of the `special_tokens_map` - is this on purpose? 

- Why does Pegasus inherit from the Reformer Tokenizer -> I don't really see what they have in common...

Would be awesome if you could take 10min to reply :-)

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes #8689, #8594, #8536

This PR refactors the Pegasus Tokenizer.

1st: It decouples the tokenizer from the Reformer Tokenizer because they don't really have much in common.
2nd: Pegasus' masked tokens are added. As stated in the [paper](https://arxiv.org/abs/1912.08777), PEGASUS has two masked tokens which are required for pre-training. Those two tokens `<mask_1>` and `<mask_2>` are added according to https://github.com/google-research/pegasus/blob/master/pegasus/ops/pretrain_parsing_ops.cc#L66 . This should solve or at least enable a solution for all three issues above.
3rd: IMO, all special tokens - which are in the case of Pegasus the tokens 2 to 104 - should be added to the `additional_special_tokens`. This is done here as well.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 1]
```diff
@@ -32,31 +37,136 @@
 }
 
 
-class PegasusTokenizer(ReformerTokenizer):
```
Comment (patrickvonplaten) [inline src/transformers/models/pegasus/tokenization_pegasus.py:35]: Pegasus has nothing to do with Reformer, so decouple it here.

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 2]
```diff
@@ -32,31 +37,136 @@
...
     max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    model_input_names = ["attention_mask"]
+
+    def __init__(
+        self,
+        vocab_file,
+        pad_token="<pad>",
+        eos_token="</s>",
+        unk_token="<unk>",
+        mask_token="<mask_2>",
```
Comment (patrickvonplaten) [inline src/transformers/models/pegasus/tokenization_pegasus.py:97]: Pegasus has two masked tokens that were previously not added to the tokenizer. There are defined as the 2nd and 3rd token according to the original implementation: https://github.com/google-research/pegasus/blob/939830367bcf411193d2b5eca2f2f90f3f9260ca/pegasus/ops/pretrain_parsing_ops.cc#L66

Comment (patrickvonplaten) [reply src/transformers/models/pegasus/tokenization_pegasus.py:97]: This resolves both this https://github.com/huggingface/transformers/issues/8536 and this https://github.com/huggingface/transformers/issues/8594 issue.

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 3]
```diff
@@ -32,31 +37,136 @@
...
+        mask_token_sent="<mask_1>",
+        additional_special_tokens=None,
+        **kwargs
+    ):
+        if additional_special_tokens is not None:
```
Comment (patrickvonplaten) [inline src/transformers/models/pegasus/tokenization_pegasus.py:102]: As Sam pointed out before tokens 2-104 were only used for pre-training. I think to add them to the `additional_special_tokens` in this case.

File: src/transformers/models/pegasus/tokenization_pegasus_fast.py [hunk 4]
```diff
@@ -39,23 +48,117 @@
...
         all_special_ids.remove(self.unk_token_id)  # <unk> is only sometimes special
-        assert all_special_ids == set([0, 1])
+
+        assert all_special_ids == set(
+            range(len(self.additional_special_tokens) + 3)
+        ), f"There should be 3 special tokens: mask_token, pad_token, and eos_token + {len(self.additional_special_tokens)} additional_special_tokens, but got {all_special_ids}"
+
         return [1 if x in all_special_ids else 0 for x in seq]
 
+    def _add_tokens(self, new_tokens: List[Union[str, AddedToken]], special_tokens=False) -> int:
```
Comment (patrickvonplaten) [inline src/transformers/models/pegasus/tokenization_pegasus_fast.py:150]: @thomwolf @LysandreJik @n1t0 - I've added this function to make some common tests pass. It's a hack - I think we have to update the rust tokenizer file for testing to correctly map the new mask tokens and the new additional tokens to the correct id s. I need some help on that I think.

File: src/transformers/models/reformer/tokenization_reformer.py [hunk 5]
```diff
@@ -56,7 +56,7 @@
 
 class ReformerTokenizer(PreTrainedTokenizer):
     """
-    Construct a Reformer tokenizer. Based on `SentencePiece <https://github.com/google/sentencepiece>`__ .
+    Construct a Reformer tokenizer. Based on SentencePiece <https://github.com/google/sentencepiece>`__ .
```
Comment (patrickvonplaten) [inline src/transformers/models/reformer/tokenization_reformer.py:59]: `
```suggestion
    Construct a Reformer tokenizer. Based on `SentencePiece <https://github.com/google/sentencepiece>`__ .
```

File: src/transformers/models/albert/tokenization_albert_fast.py [hunk 6]
```diff
@@ -71,10 +71,9 @@
 
 class AlbertTokenizerFast(PreTrainedTokenizerFast):
     """
-    Construct a "fast" ALBERT tokenizer (backed by HuggingFace's `tokenizers` library). Based on `SentencePiece
-    <https://github.com/google/sentencepiece>`__. This tokenizer inherits from
```
Comment (patrickvonplaten) [inline src/transformers/models/albert/tokenization_albert_fast.py:75]: @thomwolf @LysandreJik @n1t0 - I don't think the fast tokenizers are based on google's sentencepiece anymore, so I removed this statement from all fast tokenizers.

File: src/transformers/tokenization_utils_base.py [hunk 7]
```diff
@@ -1868,6 +1868,7 @@ def convert_added_tokens(obj: Union[AddedToken, Any]):
 
         # If there is a complementary special token map, load it
         special_tokens_map_file = resolved_vocab_files.pop("special_tokens_map_file", None)
+
```
Comment (patrickvonplaten) [inline src/transformers/tokenization_utils_base.py:1871]:
```suggestion
```

File: tests/test_tokenization_common.py [hunk 8]
```diff
@@ -2639,6 +2639,7 @@ def test_embeded_special_tokens(self):
                 tokenizer_r = self.rust_tokenizer_class.from_pretrained(pretrained_name, **kwargs)
                 tokenizer_p = self.tokenizer_class.from_pretrained(pretrained_name, **kwargs)
                 sentence = "A, <mask> AllenNLP sentence."
+
```
Comment (patrickvonplaten) [inline tests/test_tokenization_common.py:2642]:
```suggestion
```

File: tests/test_tokenization_pegasus.py [hunk 9]
```diff
@@ -26,42 +26,54 @@ def setUp(self):
         tokenizer.save_pretrained(self.tmpdirname)
 
     @cached_property
-    def pegasus_large_tokenizer(self):
+    def _large_tokenizer(self):
         return PegasusTokenizer.from_pretrained("google/pegasus-large")
 
-    @unittest.skip("add_tokens does not work yet")
```
Comment (patrickvonplaten) [inline tests/test_tokenization_pegasus.py:32]: Test seems to work, so delete this here

File: tests/test_tokenization_pegasus.py [hunk 10]
```diff
@@ -26,42 +26,54 @@ def setUp(self):
...
 
-    def test_pegasus_large_tokenizer_settings(self):
-        tokenizer = self.pegasus_large_tokenizer
+    def test_mask_tokens_rust_pegasus(self):
+        rust_tokenizer = self.rust_tokenizer_class.from_pretrained(self.tmpdirname)
+        py_tokenizer = self.tokenizer_class.from_pretrained(self.tmpdirname)
+        raw_input_str = "Let's see which <unk> is the better <unk_token_11> one <mask_1> It seems like this <mask_2> was important </s> <pad> <pad> <pad>"
+        rust_ids = rust_tokenizer([raw_input_str], return_tensors=None, add_special_tokens=False).input_ids[0]
+        py_ids = py_tokenizer([raw_input_str], return_tensors=None, add_special_tokens=False).input_ids[0]
+        # TODO: (Thom, Patrick) - this fails because the rust tokenizer does not know about the <mask_1>, <mask_2>, and those <unk_token_x> yet
```
Comment (patrickvonplaten) [inline tests/test_tokenization_pegasus.py:44]: @thomwolf @n1t0 - this test should not fail. The rust tokenizer incorrectly maps the mask tokens and <unk_token_11> at the moment.

Comment (patrickvonplaten) [reply tests/test_tokenization_pegasus.py:44]: It's now the only test that fails.

File: src/transformers/models/pegasus/tokenization_pegasus_fast.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/models/pegasus/tokenization_pegasus_fast.py:149]: Removing this hack yields the above error message. We should remove this and make the rust tokenizer behave correctly...

Comment (patrickvonplaten) [reply src/transformers/models/pegasus/tokenization_pegasus_fast.py]: removed it should be solved

File: tests/test_tokenization_pegasus.py (hunk 10 above)
Comment (patrickvonplaten) [reply tests/test_tokenization_pegasus.py:44]: Should be solved

File: src/transformers/convert_slow_tokenizer.py [hunk 11]
```diff
@@ -531,10 +531,12 @@ class BertGenerationConverter(SpmConverter):
...
     def vocab(self, proto):
         vocab = [
-            (self.original_tokenizer.pad_token, 0),
-            (self.original_tokenizer.eos_token, 0),
+            (self.original_tokenizer.pad_token, 0.0),
+            (self.original_tokenizer.eos_token, 0.0),
+            (self.original_tokenizer.mask_token_sent, 0.0),
+            (self.original_tokenizer.mask_token, 0.0),
         ]
-        vocab += [(f"unk_{i}", -100) for i in range(2, 2 + self.original_tokenizer.offset)]
```
Comment (patrickvonplaten) [inline src/transformers/convert_slow_tokenizer.py:537]: I think this was wrong previously

File: src/transformers/models/albert/tokenization_albert_fast.py (hunk 6 above)
Comment (thomwolf) [reply src/transformers/models/albert/tokenization_albert_fast.py:75]: indeed

File: tests/test_tokenization_pegasus.py (hunk 9 above)
Comment (thomwolf) [reply tests/test_tokenization_pegasus.py:32]: awesome!

**Comment by patrickvonplaten** [comment]
Checked all slow and fast tests on GPU.

**Review by thomwolf — APPROVED** [review]
Looks good! just a comment on init with added tokens

**Review by n1t0 — APPROVED** [review]
LGTM!

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 12]
```diff
@@ -32,31 +37,134 @@
...
+            assert isinstance(
+                additional_special_tokens, list
+            ), f"additional_special_tokens should be of type {type(list)}, but is {type(additional_special_tokens)}"
+
+            if mask_token_sent not in additional_special_tokens:
+                additional_special_tokens = [mask_token_sent] + additional_special_tokens
+            # fill additional tokens with ..., <unk_token_102> in case not all additional tokens are already taken
+            additional_special_tokens += [f"<unk_{i}>" for i in range(2, self.offset - len(additional_special_tokens))]
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/models/pegasus/tokenization_pegasus.py]: My plan was to only add tokens if the length is not full via `range(2, self.offset - len(additional_special_tokens)` 
-> made it cleaner now I think with a `raise ValueError` test.

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 13]
```diff
@@ -32,31 +37,148 @@
...
+        mask_token (:obj:`str`, `optional`, defaults to :obj:`"<mask_2>"`):
+            The token used for masking single token values. This is the token used when training this model with masked
+            language modeling (MLM). This is the token that the PEGASUS encoder will try to predict during pretraining.
+            It corresponds to `[MASK2]` in `PEGASUS: Pre-training with Extracted Gap-sentences for Abstractive
+            Summarization <https://arxiv.org/pdf/1912.08777.pdf>`__.
+        mask_token_sent (:obj:`str`, `optional`, defaults to :obj:`"<mask_1>"`):
+            The token used for masking whole target sentences. This is the token used when training this model with gap
+            sentences generation (GSG). This is the sentence that the PEGASUS decoder will try to predict during
+            pretraining. It corresponds to `[MASK1]` in `PEGASUS: Pre-training with Extracted Gap-sentences for
+            Abstractive Summarization <https://arxiv.org/pdf/1912.08777.pdf>`__.
```
Comment (LysandreJik) [inline src/transformers/models/pegasus/tokenization_pegasus.py:75]: Great docs here.

File: src/transformers/models/pegasus/tokenization_pegasus.py [hunk 14]
```diff
@@ -32,31 +37,148 @@
...
+            )
+            # fill additional tokens with ..., <unk_token_102> in case not all additional tokens are already taken
+            additional_special_tokens_extended += [
+                f"<unk_{i}>" for i in range(len(additional_special_tokens_extended), self.offset - 1)
+            ]
+
+            if len(set(additional_special_tokens_extended)) != len(additional_special_tokens_extended):
+                import ipdb
+
+                ipdb.set_trace()
```
Comment (LysandreJik) [inline src/transformers/models/pegasus/tokenization_pegasus.py:120]: debugging leftovers

File: src/transformers/models/pegasus/tokenization_pegasus_fast.py [hunk 15]
```diff
@@ -39,21 +48,114 @@
...
+            )
+            # fill additional tokens with ..., <unk_token_102> in case not all additional tokens are already taken
+            additional_special_tokens_extended += [
+                f"<unk_{i}>" for i in range(len(additional_special_tokens_extended), self.offset - 1)
+            ]
+
+            if len(set(additional_special_tokens_extended)) != len(additional_special_tokens_extended):
+                import ipdb
+
+                ipdb.set_trace()
```
Comment (LysandreJik) [inline src/transformers/models/pegasus/tokenization_pegasus_fast.py:128]: debugging leftovers

**Review by LysandreJik — APPROVED** [review]
You removed the links to google/sentencepiece but you kept the `Based on SentencePiece.`.

It seems to me that if we reference SentencePiece then it's good to keep a link to the library, no? Based on SentencePiece means that it's based on the library imo, maybe you wanted to say it's based on Unigram (or BPE if it's based on SentencePiece's BPE) instead?

Great changes, thanks for taking care of it!

**Comment by patrickvonplaten** [comment]
> You removed the links to google/sentencepiece but you kept the `Based on SentencePiece.`.
> 
> It seems to me that if we reference SentencePiece then it's good to keep a link to the library, no? Based on SentencePiece means that it's based on the library imo, maybe you wanted to say it's based on Unigram instead?
> 
> Great changes, thanks for taking care of it!

Good point! I also think it would be nicer to have a link to it...For now the text is always:

```
Construct a "fast" ALBERT tokenizer (backed by HuggingFace's `tokenizers` library). Based on SentencePiece.
```

=> So similar to what other FastTokenizers have written on their comments. But I agree that it could be confusing as "SentencePiece" doesn't really exist as an entity in `tokenizers` ... I think the "fast" sentencepiece tokenizers are either `BPE` or `Unigram` in tokenizers, no ? @thomwolf @n1t0 . Should I change the comments and link to their respective `tokenizers` model instead? So to 

```
Construct a "fast" ALBERT tokenizer (backed by HuggingFace's `tokenizers` library). Based on `Unigram <link to unigram in tokenizers>`__ . 
```

File: src/transformers/models/pegasus/tokenization_pegasus.py (hunk 14 above)
Comment (patrickvonplaten) [reply src/transformers/models/pegasus/tokenization_pegasus.py]: thanks

**Comment by LysandreJik** [comment]
I think your proposal makes a lot of sense!

File: docs/source/model_doc/albert.rst [hunk 16]
```diff
@@ -44,10 +44,17 @@ AlbertTokenizer
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 .. autoclass:: transformers.AlbertTokenizer
-    :members: build_inputs_with_special_tokens, get_special_tokens_mask,
+    :members: __call__, build_inputs_with_special_tokens, get_special_tokens_mask,
```
Comment (patrickvonplaten) [inline docs/source/model_doc/albert.rst:47]: The doc still needs to be nicely done. Waiting for @sgugger feedback here. Almost no model doc files have their fast tokenizers in the doc. Also, I think all tokenizers should at least show the `__call__` method in the doc as it's the most important method. Maybe @sgugger can you give me your opinion on which methods should be shown for the slow tokenizer and which for the fast tokenizer and I'll apply it to all model docs in this PR.

Comment (patrickvonplaten) [reply docs/source/model_doc/albert.rst]: Actually, the PR is getting a bit too big already - will open a new PR to add the fast tokenizer docs and merge this one.

File: src/transformers/models/pegasus/tokenization_pegasus.py (hunk 2 above)
Comment (adivekar-utexas) [reply src/transformers/models/pegasus/tokenization_pegasus.py:97]: Seems like the bos_token which is supposed to be passed into the decoder is missing?
